---
title: "BRIC vignette"
subtitle: BRIC 
abstract: BRIC integrates in-house and state-of-the-art computational tools and provides two analysis strategies, including LTMG (left-truncated mixture Gaussian model)-embedded scRNA-Seq analy-sis and co-expression gene analysis.
author: Yuzhou Chang
date: "26 February, 2020"
output:
  BiocStyle::html_document:
    number_sections: no
    toc: yes
    highlight: pygments
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Intorduction to BRIC

## General introduction 
BRIC integrates in-house and state-of-the-art computational tools and provides two analysis strategies, including bicluster-based co-expression gene analysis [(Xie, et al., 2020)](https://academic.oup.com/bioinformatics/article-abstract/36/4/1143/5567116?redirectedFrom=fulltext) and LTMG (left-truncated mixture Gaussian model)-embedded scRNA-Seq analysis [(Wan, et al., 2019)](https://academic.oup.com/nar/article/47/18/e111/5542876).

## Main function

The main idea of BRIC consists of two major strategies: 

* (i)  bicluster-based co-expression gene module analysis.
* (ii) quick mode (cluster) scRNA-Seq analysis

## Obejct structure 
The computational frame is constructed under the S4 data structure in R. The structure of `BRIC object` is: 

- -**BRIC_Object:** *name of object is called BRIC.*
  - -**raw_count:** *raw data matrix (gene in row, cell in columns, prefer using gene symbol).*
  - -**processed_count:** *normalized and imputation (default: FALSE).*
  - -**Meta_info:** *cell classification information based on LTMG and Bicluster.*
  - -**Discretization:** *discretized matrix based on qubic 1.0, which prepares for microarry and bulk RNA-Seq analysis.*
  - -**LTMG:** *LTMG slot is for storing relative results from first strategy.*
    - -**LTMG_discrete:** *Condition assigned matrix, which is generating from LTMG model.*
    - -**LTMG_BinarySingleSignal:** *binary matrix based on gene “on /off.”*
    - -**LTMG_BinaryMultisignal:** *binary matrix based on multiple condition.*
    - -**DimReduce:** *include three dimension loading score, including PCA, Tsne, and UMAP*
    - -**MarkerGene:** *Marker gene based on cell type identified by Seurat clustering method.*
    - -**Pathway:** *based on marker gene.*
    - -**tmp.Seurat:** *temporary Seurat object. In this Seurat Object, the starting matrix is LTMG signalling matrix.*
  - -**Bicluster:** **
    - -**Coreg_gene:** *co-regulatory genes are stored in this slot as dataframe; the first column is gene name and the second column is module number.*
    - -**CoCond_cell:** *co-condition cell are stored in this slot as dataframe; the first column is cell name and the second column is module number.*
    - -**MarkerGene:** *Marker gene based on cell type identified by Markov chain clustering algorithm.*
    - -**Pathway:** *genes based on co-expression gene of gene module (from Coreg_gene).*





# Requirements
## Environment

We recommend user to install BRIC on large memory (32GB) based linux operation system if user aims at analyzing bicluster-based co-expression analysis; if user aims at analyzing data by quick mode, we recommend to install BRIC on small memeory (8GB) based Windows or linux operation system; BRIC does not support MAC. 
We will assum you have the following installed:

* R (equal or greater than 3.5)

Pre-install packge
```{r pre_install, eval=FALSE, echo=TRUE}
install.packages(c('BiocManager','devtools', 'AdaptGauss', "pheatmap", 'mixtools','MCL', 'anocva', 
                   'qgraph','Rtools','ggpubr'))
BiocManager::install(c('org.Mm.eg.db','multtest', 'org.Hs.eg.db','clusterProfiler','DEsingle',
                       'DrImpute', 'scater', 'scran'))
devtools::install_github(repo = 'satijalab/seurat')
```
## Input

1. The input to BRIC is the single-cell RNA-seq expression matrix:

+ Rows correspond to genes and columns correspond to cells.
+ Expression units: the preferred expression values are RPKM/FPKM/CPM. 
+ The data file should be tab delimited.

2. BRIC also accept output files from 10X CellRanger, includinhg a folder which contains three individual files and h5 file. 

## Others

When you perform co-expression analysis, it will output several intermediate files, thus please make sure that you have write permission to the folder where BRIC is located. 

# Installation

For installation, simply type the following command in your R console, please select option 3 when R asks user to update packages:
```{r install, eval=FALSE, echo=TRUE}
devtools::install_github("BMEngineeR/BRIC", ref ="Beta",force = T)
```

# Example dataset

This tutorial run on a real dataset to illustrate the results obtained at each step.

As example, we will use Yan's data, a dataset containing 90 cells and 20,214 genes from human embryo, to conduct cell type prediction.

> Yan, L. et al. Single-cell RNA-Seq profiling of human preimplantation embryos and embryonic stem cells. Nat. Struct. Mol. Biol. 20, 1131-1139 (2013)

The original expression matrix was downloaded from <https://s3.amazonaws.com/scrnaseq-public-datasets/manual-data/yan/nsmb.2660-S2.csv>. The expression is provided as RPKM value. For convenience, we removed the space in the column names and deleted the second column(Transcript_ID). The processed data is available at <https://github.com/zy26/BRICdata/blob/master/Example%20data/Yan_RPKM>.


# 1. Input data, create BRIC object, add meta information, and preprocessing. 

BRIC can accepted 10X chromium input files, including a folder (contain gene name, cell name, and sparse matrix) and .h5 file.

## Input data

1. set working directory
```{r setwd, eval =FALSE, echo =TRUE}
setwd("d:/my_analysis/BRIC_TEST/2.Yan/")
```
2. Read from .h5 file.

```{r h5_files, eval= FALSE, echo=TRUE}
InputMatrix <- BRIC::ReadFrom10X_h5("C:/Users/cyz/Documents/BRIC/BRIC/inst/extdata/10X_5K/5k_pbmc_protein_v3_filtered_feature_bc_matrix.h5")
```

3. Read from 10X folder.

```{r folder, eval= FALSE, echo=TRUE}
InputMatrix <- BRIC::ReadFrom10X_folder("C:/Users/cyz/Documents/BRIC/BRIC/inst/extdata/10X_3K/folder_10X/")
```
4. Read from .csv or .txt file

we will use this data set as example to run the pipeline.

```{r txt, eval= TRUE, echo=TRUE}
InputMatrix <- read.table("d:/my_analysis/BRIC_TEST/2.Yan/Yan_expression.txt",header = T, row.names = 1)
```

## Add meta information

1. Create BRIC object.

```{r create_object, eval= TRUE, echo=TRUE}
object <- BRIC::CreateBRICObject(InputMatrix)
```

2. Addmeta: this step can add customized cell label by user, the format of file passing to `meta.info` is data frame of which row name should be cell ID, and column name should be cell type.    
```{r add_metadata, eval= TRUE, echo=TRUE}
object <- BRIC::AddMeta(object, meta.info = NULL)
```

3. plotmeta: plot meta information based on RNA count and Feature number. This step is for the following subset step in terms of filtering out low quality data.    
```{r plot_metadata,echo=FALSE, eval= TRUE, echo=TRUE}
BRIC::PlotMeta(object)
```


4. remove low quality data based on the previous plot.
```{r subset_data, echo=FALSE,eval= TRUE, echo=TRUE}
object <- BRIC::SubsetData(object , nFeature.upper=15000,nFeature.lower=8000,
                         Counts.upper=700000,Counts.lower=400000)
```

## Preprocesing 

User can choose perform normalization or imputation based on their need. The normalization method has two options, one is the simplist CPM normalization (default `normalization = 'LibrarySizeNormalization'`). The other is from package scran and can be opened by using parameter `normalization = 'scran'`, . The imputation method is from package DrImpute and can be opened by using parameter `IsImputation = TRUE` (default as closed).
```{r ProcessData,echo=FALSE, eval= TRUE, echo=TRUE}
object <- BRIC::ProcessData(object, normalization = "LibrarySizeNormalization", IsImputation = FALSE, seed = 123)
```


# 2. Run LTMG

The argument `Gene_use = 2000` is  top 2000 variant genes which are selected to run LTMG. For LTMG strategy, we recommend to use top 2000 gene. On the contrary, for co-expression gene analysis, we recommend to use all gene by changing `Gene_use = "all"`. 
```{r run_LTMG, echo=FALSE,eval= FALSE, echo=TRUE}
# demo only run top 500 gene for saving time.
object <- BRIC::RunLTMG(object, Gene_use = 500, seed = 123)
```

# 3. Quick mode.

## Dimension Reduction
User can use `reduction = "umap"` or `reductopm = "tsne"` to perform dimension reduction. 
```{r Run_dimReduce, eval= FALSE, echo=TRUE}
# demo only run top 500 gene for saving time.
object <- BRIC::RunDimensionReduction(object, reduction = "umap")
```

## Cluster
```{r Runclassfication, eval= FALSE, echo=TRUE}
# demo only run top 500 gene for saving time.
object <- BRIC::RunClassification(object,  k.param = 20, resolution = 0.5, algorithm = 1)
```

## Plot dimension reduction plot
```{r run_plotDim, eval= FALSE, echo=TRUE}
# demo only run top 500 gene for saving time.
BRIC::PlotDimension(object,reduction = "umap")
```
This function need user to input group that is used to plot on the figure. Input `4` means choose the Seurat0.5 group as cell label to plot.
![dim](D:/my_analysis\\BRIC_TEST\\figures\\LTMG\\dimension_selection.png)
![cluster](D:/my_analysis\\BRIC_TEST\\figures\\LTMG\\dimension_plot.png)


# 4. Biclustering based co-expression analysis

BRCI can provide biclustering function, which is based on our in-house novel algorithm, QUBIC2 (<https://github.com/maqin2001/qubic2>). Here we will show the basic biclustering usage of BRIC using a $200\times 88$ expression matrix generated from previous top 2000 variant genes. However, we recommend user should use `Gene_use = all` to generate LTMG matrix. 

## LTMG-discretized bicluster (recommend for single cell RNA-seq data)
You can type the following command to run discretization (LTMG) + biclustering directly:
```{r biclustering_basedLTMG,eval= FALSE,echo = TRUE}
object <- BRIC::RunLTMG(object, Gene_use = "all", seed = 123)
object <- BRIC::CalBinaryMultiSignal(object)
object <- BRIC::RunBicluster(object, DiscretizationModel = "LTMG",OpenDual = TRUE,
                          NumBlockOutput = 100, BlockOverlap = 0.7, BlockCellMin = 15)

```

## Quantile-discretized bicluster (recommend for bulk RNA-Seq or microarray data)

This will output several files, and among them you will find one named  `Yan_sub.txt.chars.blocks`,which contains the predicted biclusters.
Or, user may use first version discretization strategy provided by QUBIC 1.0.
```{r biclustering_basedQuantile,eval=FALSE,echo = TRUE}
object <- BRIC::RunBicluster(object, DiscretizationModel = "Quantile",OpenDual = TRUE, Extension = 0.90,
                          NumBlockOutput = 100, BlockOverlap = 0.7, BlockCellMin = 15)
```

(The default parameters in BRIC are BlockCellMin=15, BlockOverlap=0.7, Extension=0.90, NumBlockOutput=100 you may use other parameters as you like, just specify them in the argument)

## Cell type prediction based on Markove clustering

The cell type prediction of BRIC is based on the biclustering results. In short, it will construct a weighted graph based on the biclusters and then do clustering on the weighted graph. Currently, we provide two commonly used clustering method: MCL .


```{r cell_type,eval=FALSE}
object <- BRIC::FindClassBasedOnMC(object)
```
## Visualize block and network.

```{r visualize_block, eval = FALSE, echo =TRUE }
BRIC::PlotHeatmap(object ,N.block =2)
```
![blockheatmap](D:/my_analysis\\BRIC_TEST\\figures\\bicluster\\blockheatmap.png)

```{r visualize_network, eval = FALSE, echo =TRUE }
BRIC::PlotNetwork(object, Bic.index = 1, method = "pearson")
```
![network](D:/my_analysis\\BRIC_TEST\\figures\\bicluster\\network.png)

# 5. Biological interpretation. 
## Cell-type-specific marker genes
```{r CTS_marker_gene, eval = FALSE, echo =TRUE}
object <- BRIC::FindMarkers(object)
```
User need to select cell type to compare, while here we select `4:Suerat0.5` as cell type category to analyze.
![markergene](D:/my_analysis\\BRIC_TEST\\figures\\LTMG\\markergebe_selection.png)

Then BRIC will ask user choose a first group as reference, while here we select the third group (`3 : 2`) marked as cluster 2 in umap.
![markergene](D:/my_analysis\\BRIC_TEST\\figures\\LTMG\\markergene_1stGroup.png)

Then user requires to select the second group as compared object, while here user can choose either one group (`2 : 1`, `3 : 2`), or rest of all groups (`4 : rest of all`). 
![markergene](D:/my_analysis\\BRIC_TEST\\figures\\LTMG\\markergene_2stGroup.png)

After running the `Findmarker`, user can find table in `object@LTMG@MarkerGene` if using quick mode or find table in `object@BiCluster@MarkerGene`.

## Cell-type-specific pathways.
The first pathway analysis is based on quick mode by specifying `genes.source = "CTS"`, which means cell-type-specific marker genes; the second pathway analysis is based on genes from bicluster block. 
```{r CTS_pathways, eval = FALSE, echo =TRUE}
object <- BRIC::RunPathway(object, selected.gene.cutoff = 0.05,
                        species = "Human", database = "GO", genes.source = "CTS")
object <- BRIC::RunPathway(object ,module.number = 1, selected.gene.cutoff = 0.05,
                        species = "Human", database = "GO", genes.source = "Bicluster")
```


# sessioninfo
```{r}
sessionInfo()
```
